---
title: "enhancerHomologSearch Guide"
author: "Jianhong Ou"
bibliography: bibliography.bib
csl: nature.csl
vignette: >
  %\VignetteIndexEntry{enhancerHomologSearch Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  html_document:
    theme: simplex
    toc: true
    toc_float: true
    toc_depth: 4
---


```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  library(enhancerHomologSearch)
  library(BiocParallel)
  library(BSgenome.Drerio.UCSC.danRer10)
  library(BSgenome.Hsapiens.UCSC.hg38)
  library(BSgenome.Mmusculus.UCSC.mm10)
  library(ENCODExplorer)
  library(TxDb.Hsapiens.UCSC.hg38.knownGene)
  library(org.Hs.eg.db)
  library(TxDb.Mmusculus.UCSC.mm10.knownGene)
  library(org.Mm.eg.db)
  library(utils)
})
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

## Introduction

There is an increasing requirement for the tools to 
identify of putative mammalian orthologs to enhancers in species other 
than human and mouse, such as zebrafish, which is lacking whole genome 
comparison analysis data. Take zebrafish as an example, there are two major
methods to identify the orthologs to enhancers in human and mouse, 

1. use the whole genome comparison analysis data and conservation data[@howe2013zebrafish],

2. use spotted gar genome as bridge genome to search the orthologs[@braasch2016spotted].

Both of the methods will work well in the coding region. 
However, there is lacking of comparative data in distal regulation region 
such as enhancers and silencers. 

In 2020, Emily S. Wong et. al. provide a new
methods for identification of putative human orthologs to enhancers 
of zebrafish[@wong2020deep]. They used the method to interrogate conserved syntenic regions 
and human and mouse using candidate sponge enhancer sequences. 
First, they looked for overlap with available functional genomics
information. For example, they used mouse ENCODE data to infer enhancer activity
based on histone marks in specific tissues.
Second, they select the best-aligned region by whole genome alignment from the
candidates regions for human and mouse as orthologs.
This methods provide the possibility to search orthologs for enhancers or 
silencers even there is not genome comparative data available.

This package is modified from Wong's methods and provide the easy to used
script for researchers to quick search putative mammalian orthologs to enhancers.
The modified algorithm is:
The candidate regions were determined by ENCODE histone marks (default is 
H3K4me1) in specific tissue for humam and mouse[@pages2013package].
The mapping score were calculated by pairwise alignment between enhancer sequences 
and candidates by global sequence alignment[@pages2013package].
The Z-score were calculated from
mapping score and then converted to P-value based on two-side test from a 
normal distribution.
The candidates were filtered by p-value and distance from the TSS of 
target homologs. 
And then the top candidates from human and mouse were aligned to each other and
exported as multiple alignments with given enhancer[@bodenhofer2015msa].

## Step 1, prepare target enhancer sequences.
In this example, we will use an enhancer of hbegf gene in zebrafish.
```{r}
# load genome sequences
library(BSgenome.Drerio.UCSC.danRer10)
# define the enhancer genomic coordinates
hbegfEN <- GRanges("chr14", IRanges(6760805,	6761115))
# extract the sequences as Biostrings::DNAStringSet object
(seqEN <- getSeq(BSgenome.Drerio.UCSC.danRer10, hbegfEN))
```

## Step 2, download candidate regions of enhancers from ENCODE by H3K4me1 marks
By default, the hisone marker is H3K4me1. Users can also define the markers 
by `markers` parameter in the function `getENCODEdata`. To make sure the markers
are tissue specific, we can filter the data by `biosample_name` and 
`biosample_type` parameters. 
For additional filters, please refer `?getENCODEdata`.

```{r}
# load library
library(enhancerHomologSearch)
library(BSgenome.Hsapiens.UCSC.hg38)
library(BSgenome.Mmusculus.UCSC.mm10)
library(ENCODExplorer)
# download enhancer candidates for human heart tissue
hs <- getENCODEdata(genome=Hsapiens,
                    biosample_name = "heart",
                    biosample_type = "tissue")
# download enhancer candidates for mouse heart tissue
mm <- getENCODEdata(genome=Mmusculus,
                    biosample_name = "heart",
                    biosample_type = "tissue")
```

## Step 3, get alignment score for target enhancer and candidate enhancers.
This step is time consuming step. For quick run, users can subset the data
by given genomic coordinates.
```{r}
# subset the data for test run 
# in human, the homolog HBEGF gene is located at chromosome 5
# you can try to subset the data by chromosome 5
gr <- as(seqinfo(Hsapiens), "GRanges")
hs <- subsetByOverlaps(hs, gr[seqnames(gr)=="chr5"])
# In this test run, we will only use upstream 1M and downstream 1M of homolog
# gene
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(org.Hs.eg.db)
eid <- mget("HBEGF", org.Hs.egALIAS2EG)[[1]]
g_hs <- select(TxDb.Hsapiens.UCSC.hg38.knownGene,
               keys=eid,
               columns=c("GENEID", "TXCHROM", "TXSTART", "TXEND", "TXSTRAND"),
               keytype="GENEID")
g_hs <- range(with(g_hs, GRanges(TXCHROM, IRanges(TXSTART, TXEND))))
expandGR <- function(x, ext){
  stopifnot(length(x)==1)
  start(x) <- max(1, start(x)-ext)
  end(x) <- end(x)+ext
  GenomicRanges::trim(x)
}
hs <- subsetByOverlaps(hs, expandGR(g_hs, ext=1000000))
# in mouse, the homolog Hbegf gene is located at chromosome 18
# same as above script in human, you can try to subset the data by chromosome.
gr <- as(seqinfo(Mmusculus), "GRanges")
mm <- subsetByOverlaps(mm, gr[seqnames(gr)=="chr18"])
# Here we use the subset of 1M upstream and downstream of homolog gene.
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
eid <- mget("Hbegf", org.Mm.egALIAS2EG)[[1]]
g_mm <- select(TxDb.Mmusculus.UCSC.mm10.knownGene,
               keys=eid,
               columns=c("GENEID", "TXCHROM", "TXSTART", "TXEND", "TXSTRAND"),
               keytype="GENEID")
g_mm <- range(with(g_mm, GRanges(TXCHROM, IRanges(TXSTART, TXEND))))
mm <- subsetByOverlaps(mm, expandGR(g_mm, ext=1000000))

# use parallel computing to speed up.
library(BiocParallel)
bpparam <- MulticoreParam(tasks=200, progressbar=TRUE)
aln_hs <- alignmentOne(seqEN, hs, bpparam=bpparam)
aln_mm <- alignmentOne(seqEN, mm, bpparam=bpparam)
```

## Step 4, filter the candidate regions.

Here we will filter the candidate regions more than 5K from TSS of homolog 
but within 100K from the gene body. The candidates will be also filtered by 
p-value.

```{r}
# Step4
ext <- 100000
aln_hs <- subsetByOverlaps(aln_hs, ranges = expandGR(g_hs, ext=ext))
## filter by distance
aln_hs$peaks$distance <- distance(aln_hs$peaks, g_hs, ignore.strand=TRUE)
aln_hs <- subset(aln_hs, pval<0.05 & distance >5000)


aln_mm <- subsetByOverlaps(aln_mm, ranges = expandGR(g_mm, ext=ext))
## filter by distance
aln_mm$peaks$distance <- distance(aln_mm$peaks, g_mm, ignore.strand=TRUE)
aln_mm <- subset(aln_mm, pval<0.05 & distance >5000)
```

## Step 5, export the multiple alignments in order.

The selected candidates will be aligned cross human and mouse and then output
as phylip multiple alignment file in text format.

```{r}
al <- alignment(seqEN, list(human=aln_hs, mouse=aln_mm),
                method="ClustalOmega", order="input")
saveAlignments(al)
readLines("aln1.phylip.txt")
```
## Session info

```{r}
sessionInfo()
```

